test/e2e/scheduling/predicates.go

```diff
@@ -799,8 +804,72 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
 			framework.ExpectEqual(numInNode2, expected, fmt.Sprintf("Pods are not distributed as expected on node %q", nodeNames[1]))
 		})
 	})
+
+	ginkgo.It("validates Pods with non-empty schedulerGates are blocked on scheduling [Feature:PodSchedulingReadiness] [alpha]", func() {
+		podLabel := "e2e-scheduling-gates"
+		replicas := 3
+		ginkgo.By(fmt.Sprintf("Creating a Deployment with replicas=%v, carrying scheduling gates [foo bar]", replicas))
+		deployConfig := pauseDeployConfig{
+			Replicas: int32(replicas),
+			Strategy: appsv1.DeploymentStrategy{
+				Type: appsv1.RecreateDeploymentStrategyType,
+			},
+			PodConfig: pausePodConfig{
+				Name:      podLabel,
+				Namespace: ns,
+				Labels:    map[string]string{podLabel: ""},
+				SchedulingGates: []v1.PodSchedulingGate{
+					{Name: "foo"},
+					{Name: "bar"},
+				},
+			},
+		}
+		deploy := createPauseDeploy(f, deployConfig)
+
+		ginkgo.By("Expect all pods stay in pending state")
+		_, err := e2epod.WaitForNumberOfPods(cs, ns, replicas, time.Minute)
+		framework.ExpectNoError(err)
+		framework.ExpectNoError(e2epod.WaitForAllPodsSchedulingGated(cs, ns, time.Minute))
+
+		ginkgo.By("Uplift one scheduling gate")
+		clone := deploy.DeepCopy()
+		want := []v1.PodSchedulingGate{{Name: "bar"}}
+		clone.Spec.Template.Spec.SchedulingGates = want
+		deploy, err = patchDeployment(cs, deploy, clone)
```

Comment:
this is changing the deployment, not the pods, so it will practically delete the old pods and generate new ones with a single gate, is this really how we want to test this? we should update the pods themselves so the test reflects better the expected pattern for using this feature, right?

Reply:
> so it will practically delete the old pods and generate new ones with a single gate

correct.

> we should update the pods themselves so the test reflects better the expected pattern for using this feature

That's a bit tricky. What's the typical workflow of having a CR controller managing the corresponding Pods upon rollout? Say does the CR controller usually recreate pods, or update the pods in-place?

Reply:
> What's the typical workflow of having a CR controller managing the corresponding Pods upon rollout?

But the feature is about managing the pods themselves. The controller (like a quota controller) will have a dependency on the pods API only, and typically doesn't even know which application created those pods, and doesn't know how to update the pod template in that application API (think of CRDs for example). If it does, then the better approach is preventing the creation of the pods in the first place.

Reply:
> Say does the CR controller usually recreate pods, or update the pods in-place?

The controller managing the schedulingGates (like a quota controller) will update the pods in place, that is why the field is mutable for removal.

Reply:
The feature is indeed about the pods. It's just in the POV of e2e test, I want to come up with a test close to user's e2e workflow: e.g., does the controller update PodTemplate or not. But I agree recreating Pods doesn't quite qualify the e2e coverage. I will tweak the test to update plain pods.

Reply:
>  I want to come up with a test close to user's e2e workflow:

Yeah, that is why I am suggesting to change the test so we update the existing pods. I can't think of a case where a controller will want to update the podTemplate; as I mentioned, those controllers don't even know which application created the pods. Moreover, the management happens at the pod level, for example if a quota controller decides that a pod can be admitted doesn't mean that all other replicas can (which what will happen if it updates the template in the application API).

Reply:
Updated the e2e test:

- initially create a replicaset with 3 replicas
- patch 3 pods (instead of PodTemplate) iteratively in 2 rounds - [foo bar] -> [foo] -> []